java/src/main/java/ai/rapids/cudf/ColumnView.java

```diff
@@ -0,0 +1,3054 @@
+/*
+ *
+ *  Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static ai.rapids.cudf.HostColumnVector.OFFSET_SIZE;
+
+public class ColumnView implements AutoCloseable, BinaryOperable {
+
+  public static final long UNKNOWN_NULL_COUNT = -1;
+
+  protected long columnHandle;
+  protected long viewHandle = 0;
+  protected DType type;
+  protected long rows;
+  protected Optional<Long> nullCount = Optional.empty();
+
+  public ColumnView(DType type, long rows, Optional<Long> nullCount,
+                    BaseDeviceMemoryBuffer dataBuffer, BaseDeviceMemoryBuffer validityBuffer,
+                    BaseDeviceMemoryBuffer offsetBuffer, List<ColumnView.NestedColumnVector> nestedColumnVectors) {
+    if (type != DType.STRING && type != DType.LIST) {
+      assert offsetBuffer == null : "offsets are only supported for STRING, LISTS";
+    }
+    this.rows = rows;
+    this.nullCount = nullCount;
+    this.type = type;
+
+    List<BaseDeviceMemoryBuffer> toClose = new ArrayList<>();
+    long[] childHandles = new long[nestedColumnVectors.size()];
+    for (ColumnView.NestedColumnVector ncv : nestedColumnVectors) {
+      toClose.addAll(ncv.getBuffersToClose());
+    }
+    for (int i = 0; i < nestedColumnVectors.size(); i++) {
+      childHandles[i] = nestedColumnVectors.get(i).getViewHandle();
+    }
+    assert !nullCount.isPresent() || nullCount.get() <= Integer.MAX_VALUE;
+    int nc = nullCount.orElse(UNKNOWN_NULL_COUNT).intValue();
+    if (rows == 0  && !type.isNestedType()) {
+      this.columnHandle = ColumnVector.makeEmptyCudfColumn(type.typeId.getNativeId(), type.getScale());
+    } else {
+      long cd = dataBuffer == null ? 0 : dataBuffer.address;
+      long cdSize = dataBuffer == null ? 0 : dataBuffer.length;
+      long od = offsetBuffer == null ? 0 : offsetBuffer.address;
+      long vd = validityBuffer == null ? 0 : validityBuffer.address;
+      this.viewHandle = makeCudfColumnView(type.typeId.getNativeId(), type.getScale(), cd, cdSize,
+          od, vd, nc, (int)rows, childHandles) ;
+    }
+  }
+
+  //TODO can we make this better
+  protected ColumnView(long address, boolean isColumnView) {
```

Comment:
I am not terribly happy with passing a boolean but Java constructor constraints and passing of longs is what drove this change. I could build a ColumnView and instantiate viewHandle later but it just open doors for bugs and half way initialization of values we really care about. Happy to be corrected if there is a better way to do this.

Reply:
IMO `ColumnView` should never know about a column handle, only view handles.  I would expect the `ColumnVector` constructor that takes a column handle argument to call a static function to generate the column view handle from a column handle and pass that to the `ColumnView` constructor, something like this:
```java
public ColumnVector(long columnHandle) {
  super(viewHandleFromColumnHandle(columnHandle));
}

private static long viewHandleFromColumnHandle(long columnHandle) {
  ... uses a native method to generate a view handle with error handling to close column handle if necessary ...
}
```

Reply:
Agree. Will update